pints/_mcmc/_result.py

```diff
@@ -0,0 +1,157 @@
+#
+# MCMC result object
+#
+# This file is part of PINTS.
+#  Copyright (c) 2017, University of Oxford.
+#  For licensing information, see the LICENSE file distributed with the PINTS
+#  software package.
+#
+# Some code in this file was adapted from Myokit (see http://myokit.org)
+#
+from __future__ import absolute_import, division
+from __future__ import print_function, unicode_literals
+import pints._diagnostics as diagnostics
+import numpy as np
+from tabulate import tabulate
+
+
+class McmcResultObject(object):
+    """
+    Class that represents the result of running MCMC and
+    calculates key parameter summaries which are printed to the
+    console, including the posterior mean, standard deviation,
+    quantiles, rhat and effective sample size
+    """
+
+    def __init__(self, chains, time):
+        self._chains = chains
+        stacked = np.vstack(chains)
+        self._time = time
+        self._mean = np.mean(stacked, axis=0)
+        self._std = np.std(stacked, axis=0)
+        self._quantiles = np.percentile(stacked, [2.5, 25, 50,
+                                                        75, 97.5], axis=0)
+        self._ess = diagnostics.effective_sample_size(stacked)
+        self._ess_per_second = self._ess / time
+        self._num_params = stacked.shape[1]
+        self._num_chains = len(chains)
+
+        # If there is more than 1 chain calculate rhat
+        # otherwise return NA
+        if self._num_chains > 1:
+            self._rhat = diagnostics.rhat_all_params(chains)
```

Comment:
Perhaps move this to make_summary() as well?

Reply:
Good points Martin and Chon! Happy to move these calculations to make_summary; agreed that this should only be done if requested by the user. I will make these changes and make MCMC routines return these objects by default.

@martinjrobins @MichaelClerx @chonlei -- how do you feel about the extra dependency that this functionality requires (it needs a Python package called "tabulate")? It's a lightweight package that does the formatting of the table when we print the object. I'm tempted for now to include it as a core dependency but, later on, someone/me can copy its action using core Python code. Just don't want to get too caught up on dependency stuff now!